evm/combine-signatures.js

```diff
@@ -0,0 +1,273 @@
+            );
+
+            const data = defaultAbiCoder.encode(
+                ['uint256', 'bytes32[]', 'string[]', 'bytes[]'],
+                [chainId, [commandID], [command], [params]],
+            );
+
+            const dataHash = hashMessage(arrayify(keccak256(data)));
+
+            printInfo('Original bytes message (pre-hash)', data);
+            printInfo('Message hash for validators to sign', dataHash);
+
+            break;
+        }
+
+        case 'constructBatch': {
+            const { message } = options;
+
+            validateParameters({ isValidCalldata: { message } });
+
+            const {
+                addresses: validatorAddresses,
+                weights,
+                threshold,
+                keyID: expectedKeyId,
+            } = await getEVMAddresses(config, chain.id, options);
+
+            const validatorWeights = {};
+
+            validatorAddresses.forEach((address, index) => {
+                validatorWeights[address.toLowerCase()] = weights[index];
+            });
+
+            const signatures = readSignatures();
+
+            const sortedSignatures = signatures.sort((a, b) => {
+                const addressA = getAddressFromPublicKey(`0x${a.pub_key}`).toLowerCase();
+                const addressB = getAddressFromPublicKey(`0x${b.pub_key}`).toLowerCase();
+                return addressA.localeCompare(addressB);
+            });
+
+            const batchSignatures = [];
+
+            let totalWeight = 0;
+
+            const expectedMessageHash = hashMessage(arrayify(keccak256(message)));
+
+            for (const signatureJSON of sortedSignatures) {
+                const keyId = signatureJSON.key_id;
+                const msgHash = `0x${signatureJSON.msg_hash}`;
+                const pubKey = `0x${signatureJSON.pub_key}`;
+                const signature = `0x${signatureJSON.signature}`;
+
+                validateParameters({
+                    isNonEmptyString: { keyId },
+                    isKeccak256Hash: { msgHash },
+                    isValidCalldata: { pubKey, signature },
+                });
+
+                if (expectedKeyId !== keyId) {
+                    printError('Signature contains invalid key_id', keyId);
+                    return;
+                }
+
+                if (msgHash.toLowerCase() !== expectedMessageHash.toLowerCase()) {
+                    printError('Message hash does not equal expected message hash', msgHash);
+                    return;
+                }
+
+                const validatorAddress = getAddressFromPublicKey(pubKey);
+
+                // const signer = verifyMessage(msgHash, signature);
+
+                // if (signer.toLowerCase() !== validatorAddress.toLowerCase()) {
+                //     printError('Signature is invalid for the given validator address', validatorAddress);
+                //     return;
+                // }
```

Comment:
I want to add this sanity check but the way I was trying here kept failing even while `auth.validateProof` returns true, I assume that the function I am using to retrieve the original signer address is wrong, if you have any ideas @milapsheth @re1ro @blockchainguyy please let me know :)

Reply:
Works with `ethers.utils.recoverAddress`